drivers: serial: nrf: Adding UARTE driver for the nRFx family

jakub-uC opened the pull request:
1. Added support for two instances of UARTE.

2. Kconfig.nrfx is capable to configure UART and UARTE driver.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,89 @@ config UART_0_NRF_RTS_PIN
+	int "Size of RAM buffer"
+	default 32
+	help
+	  SPIM peripherals cannot transmit data directly from flash. Therefore,
```
SPIM?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+typedef struct
+{
+	NRF_UARTE_Type * p_reg;        /* Pointer to registers structure */
+	uint8_t          drv_inst_idx; /* Driver instance index */
```
What is the purpose of this? A value is assigned to this field but it is never referenced.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+	uarte_nrfx_instance_t uarte;	/* Instance address and number. */
+	nrf_uarte_hwfc_t      hwfc;	/* Flow control configuration. */
+	nrf_uarte_parity_t    parity;	/* Parity configuration. */
+	nrf_uarte_baudrate_t  baudrate;	/* Baudrate. */
```
What is the point in storing all these parameters that you only use at initialization?
Actually, you only need `max_chunk_len` and `uarte.p_reg`.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+*/
+static void uarte_nrfx_isr(void *arg)
+{
+	__ASSERT(dev, "NULL pointer");
```
dev?

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+	__ASSERT(dev, "NULL pointer");
+
+	int			  err;
+	struct device		  *gpio_dev;
```
This alignment is weird.

anangl reviewed (changes requested):
A few comments just for the beginning. There will be more...

JoeHut commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,89 @@ config UART_0_NRF_RTS_PIN
+	int "Size of RAM buffer"
+	default 32
+	help
+	  SPIM peripherals cannot transmit data directly from flash. Therefore,
```
Looks like a copy-paste failure from the nrfx SPI driver.

jarz-nordic replied:
fixed

JoeHut commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+} uarte_nrfx_data_t;
+
+typedef struct
+{
```
I think the style guide only has opening braces on the next line for functions.

jarz-nordic replied:
fixed

JoeHut commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+
+
+	if (data->cb)
+	{
```
see above.

jarz-nordic replied:
fixed

JoeHut commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,644 @@
+{
+	__ASSERT(dev, "NULL pointer");
+
+	struct device 		*dev = arg;
```
checkpatch does not like a mix of tabs and spaces I think.

jarz-nordic replied:
fixed

jarz-nordic commented:
This PR was inspired by @BertoldVdb-byteflies

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,651 @@
+			IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_UARTE##idx),      \
+				    CONFIG_UART_##idx##_IRQ_PRI,              \
+				    uarte_nrfx_isr,                           \
+				    DEVICE_GET(uart_nrfx_uarte##idx),         \
```
Since you don't use anything else from `dev` in the IRQ handler, you could pass `&uarte_##idx##_data` here and simplify the code (you'll get rid of `DEVICE_DECLARE`, too).

jarz-nordic replied:
Cannot do it as callback function requires devive structure.

anangl replied:
Right. My mistake.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,651 @@
+	#endif
+	#ifdef CONFIG_UART_0_NRF_FLOW_CONTROL
+		#undef CONFIG_UART_0_NRF_FLOW_CONTROL
+		#define CONFIG_UART_0_NRF_FLOW_CONTROL	NRF_UARTE_HWFC_ENABLED
```
What about something like this?
```
#ifdef CONFIG_UART_0_NRF_FLOW_CONTROL
	#define UART_0_NRF_HWFC_CONFIG                \
		.pselcts = CONFIG_UART_0_NRF_CTS_PIN, \
		.pselrts = CONFIG_UART_0_NRF_RTS_PIN, \
		.hwfc = NRF_UARTE_HWFC_ENABLED,
#else
	#define UART_0_NRF_HWFC_CONFIG                  \
		.pselcts = NRF_UARTE_PSEL_DISCONNECTED, \
		.pselrts = NRF_UARTE_PSEL_DISCONNECTED, \
		.hwfc = NRF_UARTE_HWFC_DISABLED,
#endif /* CONFIG_UART_0_NRF_FLOW_CONTROL */
```
Or maybe even:
```
#define UART_NRF_HWFC_ENABLED(idx)                  \
	.pselcts = CONFIG_UART_##idx##_NRF_CTS_PIN, \
	.pselrts = CONFIG_UART_##idx##_NRF_RTS_PIN, \
	.hwfc = NRF_UARTE_HWFC_ENABLED,
#define UART_NRF_HWFC_DISABLED                  \
	.pselcts = NRF_UARTE_PSEL_DISCONNECTED, \
	.pselrts = NRF_UARTE_PSEL_DISCONNECTED, \
	.hwfc = NRF_UARTE_HWFC_DISABLED,
```
and then:
```
#ifdef CONFIG_UART_0_NRF_FLOW_CONTROL
	#define UART_0_NRF_HWFC_CONFIG  UART_NRF_HWFC_ENABLED(0)
#else
	...
```

jarz-nordic replied:
done

Commit:
drivers: serial: nrf: Adding UARTE driver for the nRFx family

1. Added support for two instances of UARTE.

2. Kconfig.nrfx is capable to configure UART and UARTE driver.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jarz-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -272,11 +272,11 @@ static int uart_nrfx_irq_is_pending(struct device *dev)
 		(nrf_uart_int_enable_check(NRF_UART0,
 					   NRF_UART_INT_MASK_RXDRDY) &&
-		uart_nrfx_irq_rx_ready(dev)));
+		 uart_nrfx_irq_rx_ready(dev)));
```
fixed

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,651 @@
+	#ifndef CONFIG_UART_0_INTERRUPT_DRIVEN
+		#define CONFIG_UART_0_INTERRUPT_DRIVEN	(0u)
+	#else
+		#undef CONFIG_UART_0_INTERRUPT_DRIVEN
```
done

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,651 @@
+	static int uarte_##idx##_init(struct device *dev)		      \
+	{								      \
+		const struct uarte_init_config init_config = {		      \
+		.pseltxd = CONFIG_UART_##idx##_NRF_TX_PIN,		      \
```
done

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,645 @@
+	struct device *dev = arg;
+	const struct uarte_nrfx_data *data;
+
+	data = get_dev_data(dev);
```
fixed

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,651 @@
+#define UART_NRFX_UARTE_INVALID				(0xFFFFFFFFu)
+
+#ifdef CONFIG_UART_0_NRF_UARTE
+	DEVICE_DECLARE(uart_nrfx_uarte0);
```
done

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,645 @@
+	}
+}
+#else
+static void uarte_nrfx_isr(void *arg) {}
```
Because macro: IRQ_CONNECT requires it.

anangl replied:
When interrupts are used. But in this case they are not, and there is no need to call IRQ_CONNECT at all.

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,645 @@
+			    config->hwfc);
+
+	/* Set baud rate */
+	err = baudrate_set(dev, config->baudrate);
```
comment removed

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,645 @@
+					config->pselcts);
+	}
+
+	/* Set low control and parity bit */
```
done

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,645 @@
+
+		/* Set ENDTX event by sending empty frame */
+		nrf_uarte_tx_buffer_set(uarte, &dummy_byte, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
```
To save this byte I've used `interrupts_active` variable.
I would like to avoid passing NULL pointer. According to spec:

"If the TXD.PTR and the RXD.PTR are not pointing to the Data RAM region, an EasyDMA transfer may result in a HardFault or RAM corruption."

anangl replied:
But `interrupts_active` is a local variable and the compiler may use a core register for storing it. By using a pointer to this variable you are forcing it to be stored on the stack. Thus, this way you are saving one byte in RAM but wasting a couple in flash.
I may be wrong, but the statement from the specification refers primarily to the situation when the transfer is actually performed, i.e. the pointed memory is accessed. But here we've got a fake transfer and no memory access is needed, so `NULL` should work.
Anyway, if you want to use a real Data RAM pointer here, use `data` or `&data->rx_data`. And add a proper comment, please, so no one wonders why such pointer is passed as the parameter.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,651 @@
+		#define CONFIG_UART_0_INTERRUPT_DRIVEN	(1u)
+	#endif /* UART_0_INTERRUPT_DRIVEN */
+	#ifndef CONFIG_UART_0_IRQ_PRI
+		#define CONFIG_UART_0_IRQ_PRI		UART_NRFX_UARTE_INVALID
```
^^

jarz-nordic replied:
fixed

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,647 @@
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	static struct uarte_nrfx_data uarte_##idx##_data;		      \
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	      \
+		.max_chunk_len = (1 << UARTE##idx##_EASYDMA_MAXCNT_SIZE) - 1, \
```
^^

jarz-nordic replied:
fixed

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,598 @@
+		return 0;
+	}
+
+	if (len > config->max_chunk_len) {
```
And what about `CONFIG_UART_NRFX_RAM_BUFFER_SIZE`?

jarz-nordic replied:
This is point of discussion. It might happen that CONFIG_UART_NRFX_RAM_BUFFER_SIZE will exceed max allowed buffer. In such case using `max_chunk_len` will allow to use maxim allowed buffer size, and `CONFIG_UART_NRFX_RAM_BUFFER_SIZE` will use overflowed value.

jarz-nordic replied:
I will add checking `CONFIG_UART_NRFX_RAM_BUFFER_SIZE`

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,88 @@ config UART_0_NRF_RTS_PIN
+	  UART driver using UARTE peripheral, cannot transmit data directly from
+	  flash. Therefore, a buffer in RAM needs to be provided for each
+	  instance of UART driver, so that the driver can copy there a chunk of
+	  data from flash and transmit it.
```
I think this description should be updated, since in this case the buffer is used always, not only for transfers from flash.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,607 @@
+		    uarte_nrfx_isr,                     \
+		    DEVICE_GET(uart_nrfx_uarte##idx),   \
+		    0)
+#define UART_NRF_IRQ_DISABLED
```
This is needless.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,607 @@
+			.parity = UART_##idx##_NRF_PARITY_BIT,		      \
+			.baudrate = CONFIG_UART_##idx##_BAUD_RATE	      \
+		};							      \
+		if (UART_##idx##_INTERRUPT_DRIVEN) {			      \
```
You no longer need this `if`.

jarz-nordic replied:
fixed

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,607 @@
+	#else
+		#define UART_0_INTERRUPT_DRIVEN		(0u)
+		#define UART_0_INTERRUPTS_INIT()	UART_NRF_IRQ_DISABLED
+	#endif /* UART_0_INTERRUPT_DRIVEN */
```
`CONFIG_UART_0_...`
 Same story with UART 1.

anangl replied:
^^

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,607 @@
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
```
I overlooked it before. What if one uses a pin from P1?

jarz-nordic replied:
fixed, but not pretty

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,607 @@
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (!nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDTX)) {
+		return 0;
```
It seems that this check could be skipped. As API doc says:
```
 * @details This function is expected to be called from UART
 * interrupt handler (ISR), if uart_irq_tx_ready() returns true.
 * Result of calling this function not from an ISR is undefined
 * (hardware-dependent). [..]
```

jarz-nordic replied:
fixed

anangl commented:
The commit title should be corrected to start with "drivers: serial:".

barsok commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,606 @@
+	case 28800:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_28800;
+		break;
+	case 38400:
```
@jarz-nordic  The previous driver supported speeds 31250 and 56000. Are we deliberately dropping support for these speeds?

barsok replied:
@jarz-nordic I have double-checked that. The driver _nrf5 for Zephyr did not support the speeds mentioned. However, the NRFX/HAL do support and the SDK also provides support for these speeds. The following is a snippet of the SDK configuration file:

// <o> NRFX_UARTE_DEFAULT_CONFIG_BAUDRATE  - Default Baudrate
// <323584=> 1200 baud 
[...]
// <7716864=> 28800 baud 
// <8388608=> 31250 baud   <-----
// <10289152=> 38400 baud 
// <15007744=> 56000 baud   <-----
// <15400960=> 57600 baud 
[...]

jarz-nordic replied:
fixed

barsok commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,606 @@
+	 * to be empty before sending a character. However, without locking,
+	 * this introduces a rare yet possible race condition if the thread is
+	 * preempted between sending the byte and checking for completion.
+
```
This comment is misleading. The API dictates that the driver should send a character and wait until its transmitter is empty - true. There would be no race condition if there was a way to detect on the nRF52 that the transmitter indeed was empty so this alone does not lead to a race condition. 

Instead, the UART only supports events and an event flag is set when the transmitter has become empty. The race condition would be when higher-priority task cleared the flag at the end. Then the lower-priority task would never see this flag set and would stall. 

So perhaps this comment could be re-written to provide clarity?

jarz-nordic replied:
The API does not dictate that the driver should send a character and wait until its transmitter is empty. In fact it is oposite, first it shall check if transmitter is empty and next write a character to data register.
I have updated description with the information that UARTE supports only flag that transmitter become empty.

anangl replied:
What about rephrasing the first part to:
```
    /* The UART API dictates that poll_out should wait for the transmitter
     * to be empty before sending a character. However, the only way of
     * telling if the transmitter became empty in the UARTE peripheral is
     * to check if the ENDTX event for the previous transmission was set.
     * Since this event is not cleared automatically when a new transmission
     * is started, it must be cleared in software, and this leads to a rare
     * yet possible race condition if the thread is preempted right after
     * clearing the event but before sending a new character. The preempting
     * thread, if it also called poll_out, would then wait for the ENDTX
     * event that had no chance to become set.
```
Would this explain the issue better?

barsok commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,85 @@ config UART_0_NRF_RTS_PIN
+
+endif #NRF_UARTE_PERIPHERAL
 
 config NRF_UART_PERIPHERAL
```
"depends" is typically favoured over "select" in Kconfig because it does not push hard any options. In this file "NRF_UARTE_PERIPHERAL" is a software option and is the grandchild of the HAS_... option.
Would it be worth considering changing selects to:  
`config NRF_UART_PERIPHERAL`
`  bool`
`  default y`
`  depends on UART_0_NRF_UART`

This would imply this flow:  HAS UART -->  (and) UART ENABLED --> (so)  PERIPHERAL ENABLED

jarz-nordic replied:
fixed

jarz-nordic replied:
sorry, If I use this apprach I have an error:
`config NRF_UARTE_PERIPHERAL
	bool
	default "y" if (UARTE_0_NRF_UART || UARTE_1_NRF_UART) && UART_NRFX && SERIAL
	depends on (UARTE_0_NRF_UART || UARTE_1_NRF_UART) && UART_NRFX && SERIAL`

barsok commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -18,18 +18,35 @@ menuconfig UART_NRFX
 	select NRF_UART_PERIPHERAL
 	help
 	  Enable nRF UART without EasyDMA on port 0.
+
```
Would it make sense to make comments regarding the flags?
HAS_HW_NRF_UART0 --- the hardware supports UART0
UART_0_NRF_UART  --- the user selected to use that UART0
NRF_UARTE_PERIPHERAL --- the actual selection of the driver and options for the driver
-- Just to make reading of this file easier.

jarz-nordic replied:
I do not want to introduce this change. I would like to be consistent with other drivers like SPI, SPIM, TWI or TWIM.

ioannisg replied:
I think this is pretty straight-forward; the naming conventions are generic enough to be easily understood.

galak commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,606 @@
+	DEVICE_DECLARE(uart_nrfx_uarte##idx);				      \
+	static struct uarte_nrfx_data uarte_##idx##_data;		      \
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	      \
+		.uarte_regs = (NRF_UARTE_Type *)NRF_UARTE##idx		      \
```
The register address should be coming from device tree.

jarz-nordic replied:
why?

jarz-nordic replied:
So far these values for other drivers like: UART, SPI, SPIM, SPIS, TWI, TWIS are taken from Kconfig.
I want to be consistent in this PR.

I would like to put it as it is and in other commit change approach for this and all other drivers.

ioannisg replied:
I don't have a strong opinion on this, but since this is an entirely new driver, I might be worth doing it the way we plan to have this in the future, and use DTS as much as possible. Then we also draw the guidelines for migration to DTS.

jarz-nordic replied:
I am on the position we shall move it to DTS, but in the second step:
https://github.com/zephyrproject-rtos/zephyr/issues/8758

carlescufi replied:
@galak @ioannisg I just spoke to @jarz-nordic and we amended the issue and added him as an assignee with a Milestone for 1.13. I think it is reasonable to do the switch all in one go so that we do things in logical steps.

galak commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,606 @@
+	.hwfc = NRF_UARTE_HWFC_DISABLED,
+
+#define UART_NRF_IRQ_ENABLED(idx)			\
+	IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_UARTE##idx),\
```
IRQ # should come from DT.

jarz-nordic replied:
As above parameters, I wanted to be consistend, I am not questinging this request.

galak replied:
Not sure I follow what you mean.  Consistent with what?

jarz-nordic replied:
With implementation in other NRF drivers like UART or SPIM.

I just created an issue for that:
https://github.com/zephyrproject-rtos/zephyr/issues/8758
Please assign me as a resolver, this will be my next task.

galak replied:
Right, so any reason not to use DTS for this new driver?

galak replied:
Ok, I saw @carlescufi comment about doing this for all the nRF drivers in one go, so I'm ok.

carlescufi replied:
@galak only that @jarz-nordic would prefer to do things in order: first commit the driver, then switch this and all the rest to DTS. The issue is scheduled for 1.13 and it will be Jakub's next work item after this is merged.

galak reviewed (changes requested):
You need to update arch/arm/soc/nordic_nrf/nrf52/dts.fixup to support UARTE.

Also, IRQ and register addr should come from DT.

galak commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,85 @@ config UART_0_NRF_RTS_PIN
+
+if NRF_UARTE_PERIPHERAL
+
+config UART_NRFX_RAM_BUFFER_SIZE
```
where is this used?

jarz-nordic replied:
I somewhere lost this change. It shall be renamed to: `UART_NRFX_TX_BUFFER_SIZE`.
I will update it.

jarz-nordic replied:
fixed

jarz-nordic commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +87,88 @@ config UART_0_NRF_RTS_PIN
+
+if NRF_UARTE_PERIPHERAL
+
+config UART_NRFX_TX_BUFFER_SIZE
```
No, buffer size value is up to the user and it equals `UART_NRFX_TX_BUFFER_SIZE`.

If user will configure `UART_NRFX_TX_BUFFER_SIZE` higher than allowed tx buffer size by architecture, driver will limit it to max allowed by architecture value.

Maximum buffer size is limited by seleceted SoC and its value is provided by Nordic MDK. In fact this can be kept in DTS but not `UART_NRFX_TX_BUFFER_SIZE`.

galak replied:
Why isn't this buffer per UART instance?

jarz-nordic replied:
I will update that.

jarz-nordic replied:
fixed

galak commented:
I pushed #8761 to fixup the dts.fixup and handle both UART and UARTE on UART0.  This also fixes UART1 support which would have been broken.

JoeHut commented:
One problem we face is that the flow control pins can be on both GPIO port 0 and 1 on a nRF52840 which is currently not supported on the nrf UART driver. Could we consider this issue for UARTE already?

jarz-nordic commented:
@JoeHut With this PR all PINs for UARTE0 and UARTE1 (including flow control) can be set either on GPIO0 or GPIO1.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,629 @@
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+
+	if (len > UART_FIFO_TX_BUFFER_SIZE) {
```
Where is `UART_FIFO_TX_BUFFER_SIZE` defined?

jarz-nordic replied:
fixed

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,629 @@
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P1_DEV_NAME);
+#endif
+	UARTE_PIN_WRITE(gpio_dev, config->pseltxd, 1);
```
On second thought, I think it should be done in a different way. Let's get rid of this dependency on gpio devices, which brings only complications here (and currently leads to potential problems, since the gpio devices are initialized POST_KERNEL and we use their API here in a PRE_KERNEL_1 initialization routine), and use nrfx GPIO HAL directly.

jarz-nordic replied:
I agree, I have used nordic HAL functions to configure gpio.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,612 @@
+	DEVICE_DECLARE(uart_nrfx_uarte##idx);				      \
+	static u8_t uarte##idx##_tx_buffer[UART##idx##_BUFF_SIZE];	      \
+	static struct uarte_nrfx_data uarte_##idx##_data = {		      \
+		.tx_buffer = uarte##idx##_tx_buffer			      \
```
This won't compile when `CONFIG_UART_INTERRUPT_DRIVEN` is not defined.
You'll need to use more macros here.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,612 @@
+#include <hal/nrf_gpio.h>
+
+/* Maximum allowed bufer size for Easy DMA transfer */
+#define UART_MAX_TX_BUFF_SIZE		BIT_MASK(UARTE0_EASYDMA_MAXCNT_SIZE)
```
This should be calculated per instance. What about something like this:
```
#define UARTE_TX_BUFFER_SIZE(idx) \
	CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE < \
		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE) ? \
	CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE : \
		BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE)
```
and then just:
```
static u8_t uarte##idx##_tx_buffer[UARTE_TX_BUFFER_SIZE(idx)];
```
And all this buffer size stuff is no longer needed here. It would be better to move it closer to the `UART_NRF_UARTE_DEVICE` macro where it is actually used.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,612 @@
+ */
+struct uarte_nrfx_config {
+	NRF_UARTE_Type *uarte_regs; /* Instance address */
+	u16_t tx_buff_size;
```
`#ifdef CONFIG_UART_INTERRUPT_DRIVEN` to keep consistency.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,613 @@
+	.pselrts = CONFIG_UART_##idx##_NRF_RTS_PIN,	\
+	.hwfc = NRF_UARTE_HWFC_ENABLED,
+#define UARTE_NRF_HWFC_DISABLED				\
+	.pselcts = NRF_UARTE_PSEL_DISCONNECTED,		\
```
Would you mind aligning these backslashes to make the code look more tidy? Sometimes they are placed with no space after the longest line, sometimes with one (and this seems to be the best option), and sometimes with way too many whitespace characters (tabs and then spaces) before them (vide `UART_NRF_UARTE_DEVICE`).

jarz-nordic replied:
done

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,613 @@
+#define UARTE_CONFIG_NO_INT(idx)					\
+	static struct uarte_nrfx_data uarte_##idx##_data;		\
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	\
+		.uarte_regs = (NRF_UARTE_Type *)NRF_UARTE##idx		\
```
Can't you extract to macros only the parts that have to be different? The way it is done now, it is harder to maintain and more error-prone in case of modifications.

jarz-nordic replied:
done

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,613 @@
+
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	DEVICE_DECLARE(uart_nrfx_uarte##idx);				      \
+	UARTE_##idx##_CONFIG_INIT();					      \
```
This is misleading. One may be really surprised that in fact two data structures are defined here.

jarz-nordic replied:
fixed - created 2 macros

ioannisg reviewed (changes requested):
Some minor comments, from my side.

jarz-nordic commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,91 @@ config UART_0_NRF_RTS_PIN
 	  The GPIO pin to use for RTS.
 
-endif # UART_0_NRF_UART
+config UART_0_NRF_TX_BUFFER_SIZE
```
This option will appear only when UARTE is activated, so no doubt about it in my opinion.

jarz-nordic commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -18,18 +18,35 @@ menuconfig UART_NRFX
-if UART_0_NRF_UART
+if UART_0_NRF_UART || UART_0_NRF_UARTE
+
+config UART_0_INTERRUPT_DRIVEN
```
I will remove UART word from the description and it will follow other options.

jarz-nordic replied:
fixed

Commit:
drivers: serial: nrf: Adopting define for UART driver.

Change define used to select interrupt driven functions
from CONFIG_UART_INTERRUPT_DRIVEN to CONFIG_UART_0_INTERRUPT_DRIVEN.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

jarz-nordic commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -72,9 +89,98 @@ config UART_0_NRF_RTS_PIN
+	bool
+
+#flag to indicate that UARTE interrupts are active
+config UARTE_INTERRUPT_DRIVEN
```
fixed

jarz-nordic commented:
recheck

carlescufi commented:
@galak can you take another look?

MariuszSkamra reviewed (commented):
I checked this and btshell, hello_world is working on nrf52840 with UARTE enabled.
However there is build failure building for nrf52810 target, but I guess it's not part of this PR.

But the build failure comes from undeclared NORDIC_NRF_UART_40002000_CURRENT_SPEED
Looking into generated_dts_board.h
#define NORDIC_NRF_UARTE_40002000_CURRENT_SPEED         115200
but 
#define CONFIG_UART_0_BAUD_RATE         NORDIC_NRF_UART_40002000_CURRENT_SPEED

carlescufi reviewed (approved):
LGTM

jarz-nordic commented:
@MariuszSkamra : So far UARTE driver is expecting defines for UART.

When all drivers will be migrated to DT it will change. So as a temporary solution you can simply change:
`NORDIC_NRF_UARTE_40002000_CURRENT_SPEED`  to: 
`NORDIC_NRF_UART_40002000_CURRENT_SPEED`.

kl-cruz commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,630 @@
+}
+
+/** Interrupt driven interrupt update function */
+static int uarte_nrx_irq_update(struct device *dev)
```
I think it should be nrfx instead of nrx

jarz-nordic replied:
fixed

kl-cruz commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,630 @@
+
+/** Set the callback function */
+static void uarte_nrfx_irq_callback_set(struct device *dev,
+					uart_irq_callback_t cb)
```
alignment to struct

jarz-nordic replied:
it is correct in the source

kl-cruz reviewed (commented):
THe code at all looks good to me. However there are some places to improve doc (lack of end stops, alignment etc.).

jarz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,630 @@
+};
+
+struct uarte_init_config {
+	u32_t  pseltxd; /* TXD pin number */
```
I have removed dots from comments describing structure fields.
Btw. there is no common approach for this.

mike-scott commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -0,0 +1,629 @@
+
+	UART_NRF_UARTE_DEVICE(1);
+#endif /* CONFIG_UART_1_NRF_UARTE */
+
```
Extra line here.

mike-scott commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -481,3 +481,4 @@ DEVICE_AND_API_INIT(uart_nrfx_uart0,
 		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
 		    &uart_nrfx_uart_driver_api);
 
+
```
Can remove this extra line

mike-scott reviewed (changes requested):
Some extra lines at the end of these

jarz-nordic commented:
I've removed extra lines.

galak reviewed (approved):
Looks good, I think we need #8761 to go in before this.

carlescufi commented:
@galak #8761 merged

mike-scott commented:
@jarz-nordic I tried testing this last night with the following setup:

- nRF52840-DK 0.11.0 HW rev
- WNC-M14A2A LTE-M modem using https://github.com/zephyrproject-rtos/zephyr/pull/6344
- GPIO P1 PR: https://github.com/zephyrproject-rtos/zephyr/pull/8799
- DTS cleanup for UARTE: https://github.com/zephyrproject-rtos/zephyr/pull/8761
- Some random additional changes for nRF: https://hastebin.com/arehivedif.diff while debugging

I tried using the LTE-M modem pins as both GPIO P1 and P0 (past 31) and was unable to get a response.  I know this is a fairly complex setup and I see that several of those PRs were merged today.  So, I'm going to setup a more basic test using a TTL device.

Did you have a recommended setup?

jarz-nordic commented:
@mike-scott  Some comments from my side.
PR: #8799 is for UART driver, this PR concerns UARTE.

I have sucessufully tested this driver for UARTE0 and UARTE1 with following examples:
- hello world
- shell

I understand that you are trying to use driver from this PR. Can you debug if you reach function: uarte_nrfx_fifo_fill?
According to debug from your link you have:
`+#if 0
 	while (size--)  {
 		uart_poll_out(ctx->uart_dev, *buf++);
 	}
+#else
+	uart_fifo_fill(ctx->uart_dev, buf, size);
+#endif`

I am having some doubts about it. Function uart_fifo_fill shall be called from interrupt context while uart_poll_out not. It looks to me that you tested it in the same context.

mike-scott commented:
@jarz-nordic I was testing it both ways actually.  When I used uart_poll_out() I was hanging at:
https://github.com/zephyrproject-rtos/zephyr/pull/8640/files#diff-0de7d1227701470077e7117a57033444R243

And the data didn't seem to be sending anything when I used uart_fifo_fill()

mike-scott commented:
@jarz-nordic an update:
1. I needed to flip TX/RX pins (oddly)
2. And I'm using GPIO_P1 for controlling the modem pins.

I now can receive output from the power up sequence on the modem:
https://hastebin.com/orokuniloj.vbs

But apparently the "AT" command isn't being received on the modem end yet (should trigger an "OK" for the modem driver to continue).  Still debugging.

mike-scott commented:
Ok, I have it working now.  Apparently I had documented the RX, TX, CTS, RTS pins from the modem side.  So they all needed to be flipped.

Also, using uart_poll_out() is what works atm.  When I tried uart_fifo_fill(), it was only sending the "A" and and a carriage return so that I was getting back an ERROR code.

mike-scott commented:
FYI, here's a full log of successful modem communication: https://hastebin.com/vaburuluhu.txt
